task_manager/task_manager.hh

```diff
@@ -0,0 +1,229 @@
+        }
+
+        // Can be called only if the task has a parent task.
+        foreign_task_ptr& get_parent() {
+            return *_parent;
+        }
+
+        future<progress> get_progress() {
+            if (_children.empty()) {
+                return make_ready_future<progress>(_status.progress);
+            }
+
+            return map_reduce(_children, [] (auto& child) {
+                return smp::submit_to(child.get_owner_shard(), [&child] {
+                    return child->get_progress();
+                });
+            }, progress{}, [] (progress l, progress r) {
+                progress res{};
+                res.total = l.total + r.total;
+                res.completed = l.completed + r.completed;
+                return res;
+            });
+
+        }
+
+        virtual std::string get_module() const = 0;
+
+        virtual is_abortable is_abortable() const {
+            return is_abortable::no;
+        }
+
+        void set_number(int number) {
+            _status.number = number;
+        }
+
+        void set_type(std::string type) {
+            _status.type = type;
+        }
+
+        void change_state(task_state state) {
+            _status.state = state;
+        }
+
+        void set_keyspace(std::string keyspace) {
+            _status.keyspace = keyspace;
+        }
+
+        void set_table(std::string table) {
+            _status.table = table;
+        }
+
+        void set_entity(std::string entity) {
+            _status.entity = entity;
+        }
+
+        void set_total_progress(double progress) {
+            _status.progress.total = progress;
+        }
+
+        void report_completion_progress(double progress) {
+            _status.progress.completed += progress;
+        }
+
+        void start(task_manager& tm) {
+            _status.start_time = db_clock::now();
+            _status.state = task_manager::task_state::running;
+
+            (void)with_gate(tm.async_gate(), [f = done(), &tm, id = _status.id] () mutable {
+                return std::move(f).finally([&tm, id] {
+                    return sleep_abortable(tm.task_ttl(), tm.abort_source());
+                }).finally([&tm, id] {
+                    tm.unregister_task(id);
+                });
+            });
+        }
+
+        virtual task_ptr create_child(task_id uuid, foreign_task_ptr&& parent) const = 0;
+
+        future<> start_on_shards(task_manager& tm, std::vector<task_id>& uuids, task_ptr this_task) {
+            return tm.container().map([parent = seastar::make_foreign(this_task), &uuids] (task_manager& tm) mutable {
+                task_ptr task = parent->create_child(uuids[this_shard_id()], std::move(parent));
+                tm.register_task(task);
+                return seastar::make_foreign<task_ptr>(task);
+            }).then([this, &tm] (foreign_task_vector tasks) {
+                this->_children = std::move(tasks);
+                start(tm);
+            });
+        }
```

Comment:
I don't understand how this function will be used.
Better introduce it along with the use case.

My gut feeling is that this should happen in `task::impl::start()`, since it would create the per-shard tasks and link them with the parent.